### PR TITLE
afsocket: move TCP keepalive options setting after `accept()` call

### PR DESCRIPTION
**Pull request by OverOrion**
This should improve POSIX (and by extension NetBSD) compatibility as
the POSIX standard does not guarantee the inheritance of options between listening and peer sockets.
I moved the `socket_options_inet_setup_tcp_keepalive_timers()` function
after `accept()`, and created a new virtual function (`setup_peer_socket()`) in `SocketOptions`
which is redefinied in `SocketOptionsInet`.

Solves #3701

Signed-off-by: Parrag Szilárd <szilard.parrag@gmail.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
I think it's worth mentioning that the actual bug is coming from the fact that setting TCP keepalive and SO_KEEPALIVE options on the listening socket is not portable. Linux guarantees that those options will be inherited by the peer sockets, but it's not a standard.

NetBSD behaved really weirdly when we did that: it closed the listening socket after the specified TCP keepalive timeout and rendered the entire `network()` source inoperable. (Let's say the behavior was undefined because we set TCP keepalive timers on the listening socket, which is not really allowed on NetBSD. FreeBSD, however, does the same as Linux).

**Review by MrAnno — COMMENTED** [review]
LGTM.

There are some problems with the current `SocketOptions` hierarchy as it sets TCP options for UDP sockets as well, etc.

We probably need to move a few other options to the new `setup_peer_socket()` method as well, but I think it's a good start.

Please write a bugfix news entry for this PR.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — DISMISSED** [review]

File: modules/afsocket/socket-options-inet.c [hunk 1]
```diff
@@ -233,6 +232,17 @@ socket_options_inet_setup_socket(SocketOptions *s, gint fd, GSockAddr *addr, AFS
   return TRUE;
 }
 
+static gboolean
+socket_options_inet_setup_peer_socket(SocketOptions *s, gint fd, GSockAddr *addr)
+{
+  SocketOptionsInet *self = (SocketOptionsInet *) s;
+
+  socket_options_setup_peer_socket_method(s, fd, addr);
```
Comment (Kokan) [inline modules/afsocket/socket-options-inet.c:240]: Please do error handling based on this method return  value.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Review by MrAnno — APPROVED** [review]

**Pull request merged**